near-plugins-derive/src/access_controllable.rs

```diff
@@ -0,0 +1,287 @@
+            }
+
+            fn get_or_init_permissions(&self, account_id: &::near_sdk::AccountId) -> #bitflags_type {
+                match self.permissions.get(account_id) {
+                    Some(permissions) => permissions,
+                    None => <#bitflags_type>::empty(),
+                }
+            }
+
+            fn grant_role_unchecked(&mut self, role: #role_type, account_id: &::near_sdk::AccountId) -> bool {
+                let flag = <#bitflags_type>::from_bits(role.acl_permission())
+                    .expect(#ERR_PARSE_BITFLAG);
+                let mut permissions = self.get_or_init_permissions(account_id);
+
+                let is_new_grantee = !permissions.contains(flag);
+                if is_new_grantee {
+                    permissions.insert(flag);
+                    self.permissions.insert(account_id, &permissions);
+                    self.add_bearer(flag, account_id);
+
+                    let event = ::#cratename::access_controllable::events::RoleGranted {
+                        role: role.into(),
+                        by: ::near_sdk::env::predecessor_account_id(),
+                        to: account_id.clone(),
+                    };
+                    event.emit();
+                }
+
+                is_new_grantee
+            }
+
+            fn has_role(&self, role: #role_type, account_id: &::near_sdk::AccountId) -> bool {
+                match self.permissions.get(account_id) {
+                    Some(permissions) => {
+                        let flag = <#bitflags_type>::from_bits(role.acl_permission())
+                            .expect(#ERR_PARSE_BITFLAG);
+                        permissions.contains(flag)
+                    }
+                    None => false,
+                }
+            }
+
+            fn has_any_role(
+                &self, roles: Vec<#role_type>,
+                account_id: &::near_sdk::AccountId
+            ) -> bool {
+                let permissions: Vec<#bitflags_type> = roles
+                    .iter()
+                    .map(|role| {
+                        <#bitflags_type>::from_bits(role.acl_permission())
+                            .expect(#ERR_PARSE_BITFLAG)
+                    })
+                    .collect();
+                let target = permissions.iter().fold(
+                    <#bitflags_type>::empty(),
+                    |acc, &x| acc | x,
+                );
+                self.has_any_permission(target, account_id)
+            }
+
+            fn has_any_permission(&self, target: #bitflags_type, account_id: &::near_sdk::AccountId) -> bool {
+                let permissions = self.get_or_init_permissions(account_id);
+                target.intersects(permissions)
+            }
+
+            /// Adds `account_id` to the set of `permission` bearers.
+            fn add_bearer(&mut self, permission: #bitflags_type, account_id: &::near_sdk::AccountId) {
```

Comment:
I want to highlight that at this point, we assume that `permission` will contain **exactly** one bit active. I checked, and this is the case for the only usage of `add_bearer,` but the function is exposed publicly (i.e, developers can call `self.acl.add_bearer(...)` themselves. We should consider adding this check.

```
fn is_power_of_2(num: u128) -> bool {
   num > 0 && ((num & (num - 1)) == 0)
}
```

Reply:
Good point, added this check in 790824f. There's even [`is_power_of_two()`](https://doc.rust-lang.org/stable/std/primitive.u128.html#method.is_power_of_two).